### PR TITLE
Core 110 - CharClassPattern.toString with unit-tests

### PR DESCRIPTION
**Pull request by ys-pyrofex**

### CODE REVIEW

File: storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala [hunk 1]
```diff
@@ -444,6 +444,103 @@ final case class CharClassPattern(charSet: Set[Char], negateCharSet: Boolean = f
...
+      val finalStrings = if (pending.isEmpty) {
+        strings
+      } else {
+        listToString(pending) :: strings
+      }
+
+      finalStrings.reverse.mkString
+    }
+
+    if (charSet.isEmpty) {
```
Comment (DoSOfRedRiver) [inline storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala]: One trick you can do with such if clauses is to use pattern-matching instead:
```scala
charSet.toList match {
  case Nil => "Nil"
  case h :: Nil => "Head"
  case h :: t => "Head and tail"
}
```
Maybe you'll found it more expressive.

File: storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala [hunk 2]
```diff
@@ -444,6 +444,103 @@ final case class CharClassPattern(charSet: Set[Char], negateCharSet: Boolean = f
...
+        //one day Unicode consortium will add 3-chars surrogate, lets handle it today
+        surrogatePair.foldLeft("")((str, ch) => str + "\\u%04X".format(ch.toInt))
+      }
+
+    def unknownSetToString: String = {
+      def listToString(lst: List[Char]): String =
+        lst.size match {
+          case 0 => ""
+          case 1 => singleCharToString(lst.head)
+          case x if x > 3 => {
```
Comment (tymm) [inline storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala]: You don't need the curly brackets here.

File: storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala [hunk 3]
```diff
@@ -444,6 +444,103 @@ final case class CharClassPattern(charSet: Set[Char], negateCharSet: Boolean = f
...
+            //sequence of 4 chars or more "abcd" -> "a-d"
+            val startChar = singleCharToString(lst.last)
+            val endChar = singleCharToString(lst.head)
+            s"$startChar-$endChar"
+          }
+          case _ => {
```
Comment (tymm) [inline storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala]: See previous comment.

File: storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala [hunk 4]
```diff
@@ -444,6 +444,103 @@ final case class CharClassPattern(charSet: Set[Char], negateCharSet: Boolean = f
...
+            //sequence like "abc", no sense to convert to "a-c"
+            lst.map(c => singleCharToString(c)).mkString
```
Comment (tymm) [inline storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala]: `lst.map(singleCharToString).mkString`.

File: storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala [hunk 5]
```diff
@@ -444,6 +444,103 @@ final case class CharClassPattern(charSet: Set[Char], negateCharSet: Boolean = f
...
+          }
+        }
+
+      val sortedSet = charSet.toList.sorted
+      val (strings, pending) = sortedSet.foldLeft((Nil: List[String], Nil: List[Char])) {
+        case ((strings, pending), nextChar) => {
```
Comment (tymm) [inline storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala]: Unnecessary curly brackets.

File: storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala (hunk 1 above)
Comment (ys-pyrofex) [reply storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala]: Thanks for the tip. Finally decided to leave as is here due to possible performance hit (Set,toList conversion). Will definitely use this feature in future.

File: storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala (hunk 2 above)
Comment (ys-pyrofex) [reply storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala]: Removed, thanks.

File: storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala (hunk 3 above)
Comment (ys-pyrofex) [reply storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala]: Removed.

File: storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala (hunk 4 above)
Comment (ys-pyrofex) [reply storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala]: Updated, thanks for the tip.

File: storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (ys-pyrofex) [reply storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala]: Updated, thanks.

**Pull request merged**